### PR TITLE
fully integrate ruff for formatting and linting

### PR DESCRIPTION
**Pull request by Kircheneer**
We had already introduced ruff as a replacement for pydocstyle. This this commit, we also replace flake8 and black with it.

It was decided not to remove bandit and pylint because they still have rules that are unsupported in ruff. Furthermore, all bandit and pylint rules will remain active because the performance improvement was tiny. The following script can theoretically extract all pylint messages that are unsupported in ruff:

```python
from bs4 import BeautifulSoup
import requests

response = requests.get("https://github.com/astral-sh/ruff/issues/970#issue-1469044346")

soup = BeautifulSoup(response.text, "html.parser")

for li in soup.find_all("li", class_="task-list-item"):
    if "checked" in li.find("input").attrs:
        continue
    print(li.find_all("code")[0].get_text(), end="")
    print(",")
```

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -138,18 +117,15 @@ ignore = [
...
 
 [tool.ruff.lint.per-file-ignores]
-"nautobot_dev_example/migrations/*" = [
-    "D",  # pydocstyle
-]
-"nautobot_dev_example/tests/*" = [
-    "D",  # pydocstyle
-]
+"nautobot_dev_example/migrations/*" = ["D", "F", "E", "W"]
+"nautobot_dev_example/tests/*" = ["D", "S"]
```
Comment (snaselj) [inline pyproject.toml:128]: I prefer to keep each item on the single line, to have better diffs.

Overall, would be nice (not in this PR) to reformat all lists in `pyproject.toml` to use this format.

File: tasks.py [hunk 2]
```diff
@@ -758,7 +735,7 @@ def check_migrations(context):
         "verbose": "Enable verbose test output.",
     }
 )
-def unittest(
+def unittest(  # noqa: PLR0913
```
Comment (snaselj) [inline tasks.py:738]: Is it possible to use short name to see what the issue is this noqa disabling?

Comment (snaselj) [inline tasks.py:738]: Is it possible to use short name to see what the issue is this noqa disabling?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -16,7 +16,7 @@ env:
   APP_NAME: "nautobot-app-dev-example"
 
 jobs:
-  black:
+  formatting:
```
Comment (snaselj) [inline .github/workflows/ci.yml:19]: Wouldn't be `ruff-format` a better name for this job? Same `ruff-lint` for the other job.

I was thinking about making separate invoke tasks for these, as both ruff actions are actually different ruff commands with different arguments.

- `invoke ruff-format`
- `invoke ruff-lint`

Opinions?

**Review by snaselj — COMMENTED** [review]
Nice work, left a few considerations.

File: tasks.py (hunk 2 above)
Comment (Kircheneer) [reply tasks.py:738]: This is not possible today: https://github.com/astral-sh/ruff/issues/1773

Comment (Kircheneer) [reply tasks.py:738]: This is not possible today: https://github.com/astral-sh/ruff/issues/1773

File: .github/workflows/ci.yml (hunk 3 above)
Comment (glennmatthews) [reply .github/workflows/ci.yml]: In nautobot core we have a [single `invoke ruff` task](https://github.com/nautobot/nautobot/blob/develop/tasks.py#L586-L616); I'd be in favor of following that pattern, especially as the format/lint split in ruff is fairly arbitrary (for example, applying isort rules to imports is part of linting, not part of formatting??).

**Review by snaselj — DISMISSED** [review]
LGTM

File: nautobot_dev_example/navigation.py [hunk 4]
```diff
@@ -2,7 +2,6 @@
 
 from nautobot.apps.ui import NavMenuAddButton, NavMenuGroup, NavMenuItem, NavMenuTab
 
-
```
Comment (cmsirbu) [inline nautobot_dev_example/navigation.py:5]: Let the empty lines wars begin :)

File: pyproject.toml [hunk 5]
```diff
@@ -96,8 +71,8 @@ ignore=".venv"
 no-docstring-rgx="^(_|test_|Meta$)"
 
 [tool.pylint.messages_control]
-# Line length is enforced by Black, so pylint doesn't need to check it.
-# Pylint and Black disagree about how to format multi-line arrays; Black wins.
+# Line length is enforced by ruff, so pylint doesn't need to check it.
+# Pylint and ruff disagree about how to format multi-line arrays; ruff wins.
```
Comment (cmsirbu) [inline pyproject.toml:75]: This line doesn't make sense to me, as I'm not seeing anything disabled in pylint for multi-line arrays. Maybe it's a historical leftover and safe to remove?

**Review by cmsirbu — DISMISSED** [review]
LGTM!

**Comment by Kircheneer** [comment]
> Should this be managed by drift manager?

@itdependsnetworks and anybody reading here, we are trial-implementing this in this app, then backporting to the cookie and finally rolling it out via drift manager. So the answer is yes!

**Pull request merged**